codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java

```diff
@@ -0,0 +1,373 @@
+    public static final String MARSHAL_SMITHY_DOCUMENT_METHOD = "MarshalSmithyDocument";
+
+    private static final String SERVICE_SMITHY_DOCUMENT_INTERFACE = "smithyDocument";
+    private static final String IS_SMITHY_DOCUMENT_METHOD = "isSmithyDocument";
+
+
+    private final GoSettings settings;
+    private final GoDelegator delegator;
+    private final Set<DocumentShape> documentShapes = new TreeSet<>();
+    private final Model model;
+
+    public ProtocolDocumentGenerator(
+            GoSettings settings,
+            Model model,
+            GoDelegator delegator
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.delegator = delegator;
+    }
+
+    /**
+     * Get the set of document shapes for the service.
+     *
+     * @return the service's document shapes.
+     */
+    public Set<DocumentShape> getDocumentShapes() {
+        return documentShapes;
+    }
+
+    /**
+     * Returns whether the service has any document shapes.
+     *
+     * @return whether the service has one or more document types.
+     */
+    public boolean hasDocumentShapes() {
+        return getDocumentShapes().size() > 0;
+    }
+
+    /**
+     * Add a document shape to the generator.
+     *
+     * @param shape the document shape to add.
+     */
+    public void addDocumentShape(DocumentShape shape) {
+        documentShapes.add(shape);
+    }
+
+    /**
+     *
+     */
+    public void generateStandardTypes() {
+        generateNoSerdeType();
+        generateInternalDocumentInterface();
+        generateDocumentPackage();
+    }
+
+    private void generateDocumentPackage() {
+        if (!hasDocumentShapes()) {
+            return;
+        }
+
+        writeDocumentPackage("document.go", writer -> {
+            writer.write("type $L = $T", DOCUMENT_INTERFACE_NAME,
+                            getInternalDocumentSymbol(DOCUMENT_INTERFACE_NAME))
+                    .write("");
+
+            writer.openBlock("func $L(v interface{}) $T {", "}", NEW_LAZY_DOCUMENT,
+                            getDocumentSymbol(DOCUMENT_INTERFACE_NAME), () -> {
+                                writer.write("return $T(v)",
+                                        getInternalDocumentSymbol(INTERNAL_NEW_DOCUMENT_MARSHALER_FUNC));
+                            })
+                    .write("");
+        });
+    }
+
+    private void generateNoSerdeType() {
+        Symbol noSerde = SymbolUtils.createValueSymbolBuilder("NoSerde",
+                SmithyGoDependency.SMITHY_DOCUMENT).build();
+
+        delegator.useShapeWriter(settings.getService(model), writer -> {
+            writer.write("type $L = $T", NO_DOCUMENT_SERDE_TYPE_NAME, noSerde);
+        });
+
+        delegator.useFileWriter("./types/types.go", settings.getModuleName() + "/types", writer -> {
+            writer.write("type $L = $T", NO_DOCUMENT_SERDE_TYPE_NAME, noSerde);
+        });
+    }
+
+    private void generateInternalDocumentInterface() {
```

Comment:
Does each service have its own document interface?

Reply:
explained in an earlier comment above, going to resolve this in favor of that thread.